File: agentos_examples/cplusplus/invoke_skill.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <agent_msgs/ExecuteAction.h>
#include <json/json.h> // Include jsoncpp library

// Feedback callback function
void feedbackCallback(const agent_msgs::ExecuteFeedbackConstPtr &feedback)
{
  ROS_INFO("Feedback progress: %f, state: %s", feedback->progress, feedback->state.c_str());
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "invoke_skill");
  ros::NodeHandle nh;

  std::string arg_skill_name = "do_action";

  // Creating JSON args string
  Json::Value args_json;
  args_json["action_id"] = 4; // stand-up and ready-to-move
  Json::StreamWriterBuilder writer;
  std::string args = Json::writeString(writer, args_json);

  // Create the action client
  actionlib::SimpleActionClient<agent_msgs::ExecuteAction> ac("/agent_skill/" + arg_skill_name + "/execute", true);

  ROS_INFO("Waiting for action server to start...");
  if (!ac.waitForServer(ros::Duration(3.0)))
  {
    ROS_ERROR("Action server not available after waiting.");
    return 1;
  }

  // Creating a goal
  agent_msgs::ExecuteGoal goal;
  goal.invoker = "test";
  goal.invoke_priority = 25;
  goal.hold_time = 3.0;
  goal.args = args;

  // Send goal to the server with feedback callback
  try
  {
    ac.sendGoal(goal, actionlib::SimpleActionClient<agent_msgs::ExecuteAction>::SimpleDoneCallback(),
                actionlib::SimpleActionClient<agent_msgs::ExecuteAction>::SimpleActiveCallback(),
                &feedbackCallback);
  }
  catch (const ros::Exception &e)
  {
    ROS_ERROR("Failed to send goal: %s", e.what());
    return 1;
  }

  ROS_INFO("Waiting for result...");
  if (!ac.waitForResult(ros::Duration(10.0)))
  {
    ROS_ERROR("Action did not finish before the timeout.");
    return 1;
  }

  // Get and print the result
  agent_msgs::ExecuteResultConstPtr result = ac.getResult();
  ROS_INFO("Result: %d, Response: %s", result->result, result->response.c_str());

  return 0;
}

File: agentos_examples/python/invoke_skill.py
#! /usr/bin/env python3

import json
import rospy
import actionlib
from agent_msgs.msg import ExecuteAction, ExecuteGoal


def feedback_callback(feedback):
  rospy.loginfo(f"Feedback progress: {feedback.progress}, state: {feedback.state}")


if __name__ == '__main__':
  rospy.init_node('invoke_skill')

  arg_skill_name = 'do_action'
  args = json.dumps({'action_id': 4}) # stand-up and ready-to-move 

  cli = actionlib.SimpleActionClient(f'/agent_skill/{arg_skill_name}/execute', ExecuteAction)

  rospy.loginfo("Waiting for action server to start...")
  if not cli.wait_for_server(timeout=rospy.Duration(3.0)):
    rospy.logerr("Action server not available after waiting.")
    exit(1)

  goal = ExecuteGoal()
  goal.invoker = 'test'
  goal.invoke_priority = 25
  goal.hold_time = 3.0
  goal.args = args

  try:
    cli.send_goal(goal, feedback_cb=feedback_callback)
  except rospy.ROSException as e:
    rospy.logerr(f"Failed to send goal: {e}")
    exit(1)

  rospy.loginfo("Waiting for result...")
  if not cli.wait_for_result(timeout=rospy.Duration(10.0)):
    rospy.logerr("Action did not finish before the timeout.")
    exit(1)

  result = cli.get_result()
  rospy.loginfo(f"Result result: {result.result}, response: {result.response}")

